Knative Graduation DD (#1820)
cncf/toc: projects/knative/knative-graduation-dd.md

# Knative Graduation Due Diligence

- Link to [Graduation application issue](https://github.com/cncf/toc/issues/1509)
- Based on the [graduation DD template](https://github.com/cncf/toc/blob/main/operations/toc-templates/template-dd-pr-graduation.md)
- Project Repo(s): github.com/knative, github.com/knative-extensions
- Project Site: https://www.knative.dev
- Sub-Projects: Serving, Eventing, Functions, Client
- Communication: https://cloud-native.slack.com/ #knative and various channels with #knative- prefix
- Project points of contacts: steering@knative.team

## Graduation Evaluation Summary for Knative

### Criteria Evaluation

Lin Sun & Jeremy Rickard conducted the due diligence of Knative who applied for graduation. The project has completed the criteria that show its maturity at graduation. The following criteria implementations are noteworthy to call out:
- Knative is a mature project that are endorsed by multiple adopters.
- The project is not only vendor neutral but also has a very diverse set of maintainers, adopters and integrators. The steering committee is composed of members from 4 different set of organizations.

> - Knative documentation is fairly clear and easy to follow.

Author: ```suggestion
- Knative documentation is fairly clear and easy to follow. The documentation contains not only knative architecture, get started, install, upgrade, troubleshooting but also administrator topics such as [configuring HA for serving](https://knative.dev/docs/serving/config-ha/) or [configuring application security with security guard](https://knative.dev/docs/serving/app-security/security-guard-about/)
```

> - The project does an excellent job of making sure that its working group public meetings are well documented.

Author: ```suggestion
- The project does an excellent job of making sure that its working group public meetings are well documented.
- According to the most recent [CNCF project velocity report](https://docs.google.com/spreadsheets/d/116ZU_ltVkJip7G073ocULHxKhy4F1OgWjNjtZ1IPBWk/edit?gid=0#gid=0), knative is the 36th most active project in CNCF, with over 120 authors and over 4000 commits) for the past 12 months.
```

- Link to [Graduation application issue](https://github.com/cncf/toc/issues/1509)
- Based on the [graduation DD template](https://github.com/cncf/toc/blob/main/operations/toc-templates/template-dd-pr-graduation.md)
- Project Repo(s): github.com/knative, github.com/knative-extensions
- Project Site: https://www.knative.dev
- Sub-Projects: Serving, Eventing, Functions, Client
- Communication: https://cloud-native.slack.com/ #knative and various channels with #knative- prefix
- Project points of contacts: steering@knative.team

## Graduation Evaluation Summary for Knative

### Criteria Evaluation

Lin Sun & Jeremy Rickard conducted the due diligence of Knative who applied for graduation. The project has completed the criteria that show its maturity at graduation. The following criteria implementations are noteworthy to call out:
- Knative is a mature project that are endorsed by multiple adopters.
- The project is not only vendor neutral but also has a very diverse set of maintainers, adopters and integrators. The steering committee is composed of members from 4 different set of organizations.
- Knative documentation is fairly clear and easy to follow.

> - The project does an excellent job of making sure that its working group public meetings are well documented.

Reviewer: Are there architectural or technical specifics that identify this project as being mature? What exactly makes this project a mature project and what are the trends you see as to why adopters are adopting this project?

Author: Yep expanded the doc section on what I observed, adding docs related to upgrade, troubleshooting but also administrator topics such as [configuring HA for serving](https://knative.dev/docs/serving/config-ha/) or [configuring application security with security guard](https://knative.dev/docs/serving/app-security/security-guard-about/)

Author: also added project high velocity data below.  Let me know if this addresses your comments, thank you!!

TAG App Delivery - Presentation Nov 8th, 2023 ([slides](https://docs.google.com/presentation/d/1ZxsRVTYW2vdlqQqVkLTQupLIDe4Ie8aBzRy4PNDaxmY/edit#slide=id.p)). Note this was in person during KubeCon thus no recording.

- [x]  **All project metadata and resources are [vendor-neutral](https://contribute.cncf.io/maintainers/community/vendor-neutrality/).**

The project's GitHub repos and project site are aligned with the TOC expectations for vendor neutrality.

- [X] **Review and acknowledgement of expectations for [graduated](sandbox.cncf.io) projects and requirements for moving forward through the CNCF Maturity levels.**		
   - [X] We [met](https://github.com/cncf/toc/issues/1509#issuecomment-2844841621) with the project lead and went over the expectation and requirements for graduated project on 01-May-2025.

- [x] **Due Diligence Review.**

Completion of this due diligence document, resolution of concerns raised, and presented for public comment satisfies the Due Diligence Review criteria.

- [X] **Additional documentation as appropriate for project type, e.g.: installation documentation, end user documentation, reference implementation and/or code samples.**

> Knative the installation doc explained multiple ways to install clearly. Saw plenty of code samples. Saw limited production recommendation doc and asked the team to add more.

Author: ```suggestion
Knative provides a few additional types of documentation to support adopters. The installation doc provided a variety of options that is easy to understand with code samples. The project also has a production recommendations doc with some limited content and the TOC has suggested adding more to assist adopters in covering their production challenges and opportunities with Knative at scale.
```

- [X] **A number of active maintainers which is appropriate to the size and scope of the project.**

For the current cadence of changes in the project and backlog of work, the project has sufficient active maintainers to sustain its current and future momentum.

- [X] **Document a complete maintainer lifecycle process (including roles, onboarding, offboarding, and emeritus status).**

This is [documented](https://github.com/knative/community/blob/main/ROLES.md#approver) well including onboarding, offboarding and emeritus status.

- [X] **Demonstrate usage of the maintainer lifecycle with outcomes, either through the addition or replacement of maintainers as project events have required.**

From a few most recent PRs, this appears to be managed well. For example:
- Remove someone off a working group lead: https://github.com/knative/community/pull/1682 
- Add someone to a working group lead: https://github.com/knative/community/pull/1677

- [X] **Project maintainers from at least 2 organizations that demonstrates survivability.**

> Confirmed that knative has maintainers from at least 2 organizations based on the [steering committee](https://github.com/knative/community/blob/main/STEERING-COMMITTEE.md#committee-members) and [working group leads](https://github.com/knative/community/blob/main/working-groups/WORKING-GROUPS.md#working-groups) documents.

Author: ```suggestion
Confirmed that knative has maintainers from at least 2 organizations based on the [steering committee](https://github.com/knative/community/blob/main/STEERING-COMMITTEE.md#committee-members) and [working group leads](https://github.com/knative/community/blob/main/working-groups/WORKING-GROUPS.md#working-groups) documents. The [maintainer file](https://github.com/knative/community/blob/main/MAINTAINERS.md) also show a diverse set of maintainers from multiple companies (Red Hat, IBM, AWS, Bloomberg, Vmware, etc).
```

The project has a [vulnerability disclosure response policy](https://github.com/knative/community/blob/main/working-groups/security/responding.md) that outlines the Product Security Taskforce, as well as the process for fixing an issue identified as a vulnerability and general timelines for remediation.

The project also has a documented policy for [early disclosure of security vulnerabilities](https://github.com/knative/community/blob/main/working-groups/security/disclosure.md) for distributors of Knative.

- [x] **Document Security Self-Assessment.**

Knative has completed a [security self-assessment](https://github.com/cncf/tag-security/blob/main/community/assessments/projects/knative/self-assessment.md).

- [x] **Third Party Security Review.**

  - [ ] Moderate and low findings from the Third Party Security Review are planned/tracked for resolution as well as overall thematic findings, such as: improving project contribution guide providing a PR review guide to look for memory leaks and other vulnerabilities the project may be susceptible to by design or language choice ensuring adequate test coverage on all PRs.

> Knative underwent a [Third Party Security Audit](https://github.com/knative/docs/blob/main/reports/ADA-knative-security-audit-2023.pdf) by Ada Logics in 2023. The audit found 16 security issues, 15 of which were fixed with upstream patches. The remaining issue was not addressed, because it was only exploitable if fundamental security assumptions of the cluster were already broken.

Reviewer 2: filled in checkbox at #L323. 

The remaining item is not going to be fixed, updated to reflect this.

### Final Assessment

The TOC has found the project to have satisfied the criteria for Graduation.

### Criteria

## Application Process Principles

### Suggested

N/A

### Required

- [x] **Give a presentation and engage with the domain specific TAG(s) to increase awareness**

TAG Runtime - Presentation Feb 3rd, 2022 ([notes](https://docs.google.com/document/d/1k7VNetgbuDNyIs_87GLQRH2W5SLgjgOhB6pDyv89MYk/edit?tab=t.0#heading=h.wjqgcvq8889g), [video](https://www.youtube.com/watch?v=Qt--cUJOaQY&feature=youtu.be))
TAG App Delivery - Presentation Nov 8th, 2023 ([slides](https://docs.google.com/presentation/d/1ZxsRVTYW2vdlqQqVkLTQupLIDe4Ie8aBzRy4PNDaxmY/edit#slide=id.p)). Note this was in person during KubeCon thus no recording.

- [x]  **All project metadata and resources are [vendor-neutral](https://contribute.cncf.io/maintainers/community/vendor-neutrality/).**

Searched the GitHub repos, browsed through the project website, all look vendor neutral.

> - [X] **Review and acknowledgement of expectations for [graduated](sandbox.cncf.io) projects and requirements for moving forward through the CNCF Maturity levels.**		

Reviewer 3: Here is the fix for this inconsistency:
https://github.com/cncf/toc/pull/1823

### Suggested

- [x]  **Governance has continuously been iterated upon by the project as a result of their experience applying it, with the governance history demonstrating evolution of maturity alongside the project's maturity evolution.**

[Governance review for the Knative project](https://docs.google.com/document/d/1kwlizyag6EsMn9GECYSam-beSosQZhLj5Q1R_FLYQyo/edit?tab=t.0) was performed in May 2025 and it looks reasonable and healthy for the project. No major must fix issues were flagged. 

### Required

- [X] **Clear and discoverable project governance documentation.**

The TOC was able to discover the [governance file](https://github.com/knative/community/blob/main/GOVERNANCE.md) easily after suggesting the team to add one.

- [X] **Governance is up to date with actual project activities, including any meetings, elections, leadership, or approval processes.**

Based on the governance file, the TOC was able to locate the leadership team, recent meetings with notes and recording, past election results.

- [X] **Governance clearly documents [vendor-neutral](https://contribute.cncf.io/maintainers/community/vendor-neutrality/) of project direction.**

> It is clearly documented in the governance doc: https://github.com/knative/community/blob/main/GOVERNANCE.md

Author: It stated as vendor neutral in the first sentence of the https://github.com/knative/community/blob/main/GOVERNANCE.md doc. It also mentioned in the doc who are elegible for participation of SC elections of knative, which shapes the project direction. The current composition of SC looks relative vendor neutral to me as well.

- [x] **Document project goals and objectives that illustrate the project’s differentiation in the Cloud Native landscape as well as outlines how this project fulfills an outstanding need and/or solves a problem differently.**

The [Knative website](https://knative.dev/docs/) identifies the goal of the project is to be: 
_"The easiest way to build and run serverless workloads on Kubernetes."

The website identifies two main subprojects: _serving_ and _eventing_. 

> There is a good overview of the both the [serving subproject](https://github.com/knative/specs/blob/main/specs/serving/overview.md) and the [eventing subproject](https://github.com/knative/specs/tree/main/specs/eventing) in the project Github repository. The Knative Github repository includes both motivation and goals for both subprojects. This information can also be found on the project website.

Reviewer 2: applied this, thank you for the suggestion.
